File: src/main/java/phss/wsworkcars/controllers/ModeloController.java
package phss.wsworkcars.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import phss.wsworkcars.models.Marca;
import phss.wsworkcars.models.Modelo;
import phss.wsworkcars.services.MarcaService;
import phss.wsworkcars.services.ModeloService;

import java.util.List;

@RestController
public class ModeloController {

    @Autowired
    private ModeloService service;

    @Autowired
    private MarcaService marcaService;

    @GetMapping("/modelos")
    public List<Modelo> getModelos() {
        return service.getData();
    }

    @GetMapping("/getModelo/{id}")
    public Modelo getModeloById(@PathVariable long id) {
        return service.getDataById(id).orElse(null);
    }

    @GetMapping("/getModeloByName/{name}")
    public Modelo getModeloByName(@PathVariable String name) {
        return service.getModeloByName(name).orElse(null);
    }

    @GetMapping("/getModelosByMarca/{marcaId}")
    public List<Modelo> getModelosByMarca(@PathVariable long marcaId) {
        return service.getModelosByMarcaId(marcaId);
    }

    @PostMapping("/addModelo")
    public Modelo addModelo(@RequestParam long marcaId, @RequestParam String nome, @RequestParam double valor_fipe) {
        Marca marca = marcaService.getDataById(marcaId).orElse(null);
        if (marca == null) return null;

        return service.saveData(new Modelo(marca, nome, valor_fipe));
    }

    @PostMapping("/addModeloJson")
    public Modelo addModeloJson(@RequestBody Modelo modelo) {
        return service.saveData(modelo);
    }

    @PostMapping("/addModelos")
    public List<Modelo> addModelos(@RequestBody List<Modelo> modelos) {
        return service.saveDataList(modelos);
    }

    @PutMapping("/updateModelo")
    public Modelo updateModelo(@RequestBody Modelo modelo) {
        return service.updateModelo(modelo);
    }

    /**
     * Caso o sistema encontre um modelo com o @param id informado
     * o retorno será OK, caso contrário o retorno será NOT_FOUND.
     */
    @DeleteMapping("/deleteModelo/{id}")
    public HttpStatus deleteModelo(@PathVariable long id) {
        Modelo result = service.deleteDataById(id);

        if (result == null) return HttpStatus.NOT_FOUND;
        else return HttpStatus.OK;
    }

}

File: src/main/kotlin/phss/wsworkcars/controllers/ModeloController.kt
package phss.wsworkcars.controllers

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import phss.wsworkcars.models.Marca
import phss.wsworkcars.models.Modelo
import phss.wsworkcars.services.MarcaService
import phss.wsworkcars.services.ModeloService


@RestController
class ModeloController {

    @Autowired lateinit var service: ModeloService
    @Autowired lateinit var marcaService: MarcaService

    @GetMapping("/modelos")
    fun getModelos(): List<Modelo> {
        return service.getData()
    }

    @GetMapping("/getModelo/{id}")
    fun getModeloById(@PathVariable id: Long): Modelo? {
        return service.getDataById(id)
    }

    @GetMapping("/getModeloByName/{name}")
    fun getModeloByName(@PathVariable name: String): Modelo? {
        return service.getModeloByName(name)
    }

    @GetMapping("/getModelosByMarca/{marcaId}")
    fun getModelosByMarca(@PathVariable marcaId: Long): List<Modelo> {
        return service.getModelosByMarcaId(marcaId)
    }

    @PostMapping("/addModelo")
    fun addModelo(
        @RequestParam marcaId: Long,
        @RequestParam nome: String,
        @RequestParam valor_fipe: Double
    ): Modelo? {
        val marca: Marca = marcaService.getDataById(marcaId) ?: return null
        return service.saveData(Modelo(marca, nome, valor_fipe))
    }

    @PostMapping("/addModeloJson")
    fun addModeloJson(@RequestBody modelo: Modelo): Modelo {
        return service.saveData(modelo)
    }

    @PostMapping("/addModelos")
    fun addModelos(@RequestBody modelos: List<Modelo>): List<Modelo> {
        return service.saveDataList(modelos)
    }

    @PutMapping("/updateModelo")
    fun updateModelo(@RequestBody modelo: Modelo): Modelo {
        return service.updateModelo(modelo)
    }

    @DeleteMapping("/deleteModelo/{id}")
    fun deleteModelo(@PathVariable id: Long): HttpStatus {
        val result = service.deleteDataById(id)
        return if (result == null) HttpStatus.NOT_FOUND else HttpStatus.OK
    }

}